core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java

```diff
@@ -641,17 +700,49 @@ int getNumBuffered() {
      */
     InitialSearchPhase.SearchPhaseResults<SearchPhaseResult> newSearchPhaseResults(SearchRequest request, int numShards) {
         SearchSourceBuilder source = request.source();
-        if (source != null && source.aggregations() != null) {
+        boolean isScrollRequest = request.scroll() != null;
+        final boolean hasAggs = source != null && source.aggregations() != null;
+        final boolean hasTopDocs = source == null || source.size() != 0;
+
+        if (isScrollRequest == false && (hasAggs || hasTopDocs)) {
+            // no incremental reduce if scroll is used - we only hit a single shard or sometimes more...
```

Comment:
@s1monw Would you mind explaining why cannot use incremental reduce if scroll is used. This confuses me.